feat: Add transfer account authority ix

exdx opened the pull request:
This PR adds the ability to transfer account authority to another account via a feature flag. This provides the ability to migrate accounts on the platform.

exdx commented:
Not ready for review.

exdx reviewed (commented):
Pretty small set of changes if this is all we need. 

I don't see a way to set flags from the CLI -- I'm wondering if we wanted to add that? Could be a separate PR

jkbpvsc commented on programs/marginfi/src/state/marginfi_account.rs:
```diff
@@ -77,6 +78,24 @@ impl MarginfiAccount {
+    pub fn set_new_account_authority(
+        &mut self,
+        new_authority: Pubkey,
+    ) -> std::result::Result<(), MarginfiError> {
```
Signature is good, you can use a type alias we already have for this Result in `create::prelude`.
I would add to the name of the fn to `_checked` so its clear there are conditions to the change.

jkbpvsc commented on programs/marginfi/src/state/marginfi_account.rs:
```diff
@@ -77,6 +78,24 @@ impl MarginfiAccount {
+        new_authority: Pubkey,
+    ) -> std::result::Result<(), MarginfiError> {
+        // check if new account authority flag is set
+        if !self.get_flag(TRANSFER_AUTHORITY_ALLOWED_FLAG) || self.get_flag(DISABLED_FLAG) {
+            return Err(MarginfiError::IllegalAccountAuthorityTransfer);
```
Yes that is the assumption we are working with. 
Last thing your code is missing is the actual instruction itself, the account layout (and security declaration) and including the instruction in the `#[program]` in the `lib.rs`

jkbpvsc commented:
If you can also add a few integration tests that cover theses cases:
- Non owner tries to change the account authority -> fail
- Non whitelisted (no flag set) owner tries to change the authority -> fail
- Whitelisted owner tries to change the authority -> success

jkbpvsc commented on programs/marginfi/src/instructions/marginfi_account/transfer_authority.rs:
```diff
@@ -0,0 +1,50 @@
+    )]
+    pub signer: Signer<'info>,
+
+    pub new_authority: AccountLoader<'info, MarginfiAccount>,
```
The `new_authority` is not a `MarginfiAccount` account, its just an address which we will check as the signer moving forward.

Meaning we should use 
```
 pub new_authority: AccountInfo<'info>,
```

jkbpvsc commented on programs/marginfi/src/instructions/marginfi_account/transfer_authority.rs:
```diff
@@ -0,0 +1,50 @@
+) -> MarginfiResult {
+    // Gather accounts
+    let mut marginfi_account = ctx.accounts.marginfi_account.load_mut()?;
+    let new_account_authority = ctx.accounts.new_authority.load()?;
```
No need to load this as if its an account loader, bc we are only interested in the address of the account info

jkbpvsc commented on programs/marginfi/tests/marginfi_account.rs:
```diff
@@ -2410,3 +2411,195 @@ async fn lending_account_close_balance() -> anyhow::Result<()> {
+            fee_payer: todo!(),
+        }
+        .to_account_metas(None),
+        data: vec![],
```
Can you add a helper function to the MarginfiAccountFixture `try_transfer_account_authority` so we are not duplicating the logic across tests

jkbpvsc replied:
should be marginfi::instruction::MarginfiAccountSetAccountAuthority

jkbpvsc commented:
Also seems tests have failed

jkbpvsc commented on programs/marginfi/tests/marginfi_account.rs:
```diff
@@ -2410,3 +2411,195 @@ async fn lending_account_close_balance() -> anyhow::Result<()> {
+    // TODO: fix types
+    let transfer_account_authority_ix = Instruction {
+        program_id: marginfi::id(),
+        accounts: marginfi::instructions::MarginfiAccountSetAccountAuthority {
```
Shoule be marginfi::accounts::MarginfiAccountSetAccountAuthority

exdx commented:
Implemented the new functionality and it looks really good now. Running into some CI issues though:
```
error[E0080]: evaluation of constant value failed
  --> programs/marginfi/src/state/marginfi_account.rs:26:1
   |
26 | assert_struct_size!(MarginfiAccount, 2304);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempt to compute `0_usize - 1_usize`, which would overflow
   |
   = note: this error originates in the macro `const_assert` which comes from the expansion of the macro `assert_struct_size` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I get this error when building on an M1 mac but they went away on an x86 box. I don't think any struct sizes have changed? Trying to see if I changed something

exdx commented:
The issue is that the `MarginfiAccountSetAccountAuthority` accounts type is too large for the stack. It's an anchor issue. Looking into how to slim down the size of the struct by boxing accounts

jkbpvsc commented on test-utils/src/marginfi_account.rs:
```diff
@@ -625,4 +625,39 @@ impl MarginfiAccountFixture {
+
+    /// Use the client to send the transfer ix authority transaction
+    pub async fn try_transfer_account_authority(
+        &self,
```
`new_authority` should probably be a fn argument here

exdx replied:
I agree. I seem to break the test every time I add this though lol. 😅

exdx commented:
Seems like the CI problems here are across open PRs and not related to this PR's functionality.

exdx commented:
The fuzz test is broken across CI currently -- not related to the changes made in this PR.
